kmod/block.c

```diff
@@ -169,29 +171,55 @@ struct block_socket {
     spinlock_t list_lock;
     // We schedule write work as kernel is unhappy if we write from sk_write_space callback
     // write work is scheduled on ternfs_fast_wq and we yield after at most 10ms of writing
+    // we also check if socket needs to be cleaned up due to error timeout and do it from same work
     struct work_struct write_work;
-    // We schedule cleanup work on ternfs_fast_wq when we error out socket or if there was no activity for block_ops->timeout_jiffies
-    struct work_struct cleanup_work;
+    atomic_t write_work_active;
+    atomic_t refcount;
     // There could be multiple cleanup items scheduled, we only want first one to remove socket from hash, wake up waiters, remove requests
     bool terminal;
     // Read end. "sk_data_ready" callback does all the work.
     struct list_head read;
     // used for waiting on connect or waiting for socket do be fully removed
     struct completion sock_wait;
-    // we need number of waiters as last waiter needs to schedule work to free the socket (it will be terminal by then)
-    int waiters;
     // Saved callbacks
     void (*saved_state_change)(struct sock *sk);
     void (*saved_data_ready)(struct sock *sk);
     void (*saved_write_space)(struct sock *sk);
 };
 
+inline void block_socket_get(struct block_socket* socket) {
```

Comment:
Am I correct that you must hold RCU read lock to call this? If yes, I'd comment as such right above the function signature.